Fix of ignoring routing key parameter and passing just empty string.

LuboVarga opened the pull request:
Just small fix.

LuboVarga committed:
Fix of ignoring routing key parameter and passing just empty string.

LuboVarga committed:
WIP Migrated to latest gatling release plus few smaller changes. This build is compilable, but tests does not run.

Fixed all property creations with dangerous design (through gatling core is full of it) http://doc.akka.io/docs/akka/snapshot/scala/actors.html#Dangerous_Variants
Also moved props to companion objects (as recomended http://doc.akka.io/docs/akka/snapshot/scala/actors.html#Recommended_Practices)
Incremented Amqp client dependency version
Incremented scala version
Added commented out line in build.sbt file to be able to publish artifact to local maven directory (.m2/repository) instead of ivy dir. To use it, just uncomment and run sbt publish.
There is some problem in AmqpActionBuilder probably, which causes tests not working.

LuboVarga commented:
This pull request contains NOT WORKING code. It is compilable, but it seems to loop when run and fail on EOF after a few seconds. Please have a look at it, as I am not able to get what is wrong on it. There is probably some change of philosophy of gatling library and my changes does make bad use of it. I would like to have amqp support for release version of gatling.

maiha commented:
Oh, nice work for migrating from gatling-2.2.0-M3 to 2.2.0!
I'm afraid there are some typos. Could you check it, please?

thanks.

maiha commented on src/main/scala/io/gatling/amqp/action/AmqpActionBuilder.scala:
```diff
@@ -4,16 +4,19 @@ import akka.actor._
       case req: ConsumeRequest =>
-        system.actorOf(Props(new AmqpConsumeAction(req, next)), actorName("AmqpConsumeAction"))
+        //system.actorOf(AmqpPublishAction.props(req, next), "AmqpConsumeAction")
+        new AmqpPublishAction(req, next)
```
This and above comment should be `AmqpConsumeAction` ?

maiha commented on src/main/scala/io/gatling/amqp/action/AmqpPublishAction.scala:
```diff
@@ -4,14 +4,29 @@ import akka.actor._
+object AmqpPublishAction {
+  def props(req: PublishRequest, next: Action): Props = Props(classOf[AmqpPublishAction], req, next)
+
+  def props(req: ConsumeRequest, next: Action): Props = Props(classOf[AmqpPublishAction], req, next)
```
same here. `AmqpConsumeAction` ?

maiha commented on src/main/scala/io/gatling/amqp/config/AmqpVariables.scala:
```diff
@@ -29,9 +29,9 @@ trait AmqpVariables { this: AmqpProtocol =>
+    _manage = Some(system.actorOf(AmqpManage.props(statsEngine, this), "AmqpManage"))
+    _nacker = Some(system.actorOf(AmqpManage.props(statsEngine, this), "AmqpNacker"))
+    _router = Some(system.actorOf(AmqpManage.props(statsEngine, this), "AmqpRouter"))
+    _tracer = Some(system.actorOf(AmqpManage.props(statsEngine, this), "AmqpTracer"))
```
not `Manage` but `Nacker` , `Router` , `Tracer` in above lines?

maiha commented on src/main/scala/io/gatling/amqp/data/PublishRequest.scala:
```diff
@@ -30,7 +30,7 @@ object PublishRequest {
 
   def apply(exchange: Exchange, routingKey: String, bytes: Array[Byte]): PublishRequest =
-    new PublishRequest(exchange, "", props(), bytes)
+    new PublishRequest(exchange, routingKey, props(), bytes)
```
nice! :)

LuboVarga committed:
Revert "WIP Migrated to latest gatling release plus few smaller changes. This build is compilable, but tests does not run."

This reverts commit f9e8a53b8928167da6b036393f517e268ec57cd9.
Undone last commit. Going to do changes in smaller commits.

LuboVarga committed:
Changed publish settings to local maven repo.

LuboVarga committed:
Upgraded scala version to 2.11.8.

LuboVarga committed:
Improvements of props creation for actors.

Fixed all property creations with dangerous design (through gatling core is full of it) http://doc.akka.io/docs/akka/snapshot/scala/actors.html#Dangerous_Variants
Also moved props to companion objects (as recommended http://doc.akka.io/docs/akka/snapshot/scala/actors.html#Recommended_Practices)

LuboVarga committed:
Added stats directory (created by default by run script) info git ignorelist.

LuboVarga committed:
Added some documentation to run usage script.

LuboVarga committed:
WIP Migrated to latest (2.2.0) gatling release. This build is compilable, but tests does not run.

Few things was inspired/done like in http protocol.

LuboVarga commented:
I have today spent most of day doing migration again and properly, but still I am something missing. I have no more time allocated to spent on this, so I will use your original code, or jmeter. I will see, but my first tests had not work due to publish(..).consume(..) tuple not working on single connection (just want to test rpc method which communicate through rabbitmq exchange). Gatling lacks of documentation and it is not easy to make something work, if I do not see intent from code.

maiha commented:
Oh, this PR seems heavy work due to complex issues.
I'd like to simplify them by splitting into multiple issues (or PR).

issue1) fix bug about routingKey (f2b11d5) : merge now
issue2) bump scala version to 2.11.8 (90cce75) : merge soon
issue3) refactor actor props : pending (due to not essential)
issue4) bump gatling version to 2.2.0 : merge ASAP when I hava a chance
issue5) main issue about sharing connections

I'll try issue 1 and 2 today.
Or, feel free to rebase or create another PRs about them :)

LuboVarga committed:
Fix of consimer. It was not working.

LuboVarga committed:
Saving last consumed message in session.

LuboVarga committed:
Small improvement in attributes saving.

LuboVarga committed:
Added parameter to optinaly save consumed message to session.

Also done some doc addition and some default parameters for consume dsl part.

LuboVarga committed:
Make development possible through using maven. Now can be project linked in one IDE with project using this library and debugged nicely.

LuboVarga committed:
Added better info message for case, when there is already exchange present (with different attributes).

LuboVarga commented:
Thanks for bothering with my changes. I am new on github, so feel free to recommend me some tutorial about PR. I would like to send given issues as separate thing, but I do not know how. I have also done one branch called [exchangePreconditionFailed](https://github.com/maiha/gatling-amqp/compare/master...LuboVarga:exchangePreconditionFailed) on my forked repo of gatling-amqp and done some stuff there (for 2.2.0-M3 dependency version). If you think they are ok, merge them.

PS: I found out, that your plug-in is not what I need. It is a bit more asynchronous that I want (consume does not block and I need to simulate rpc through amqp, so I need that consume consumes only single message and save that message in session). I will give it one more day and try to make my own "consumeOne" thing to make my scenario possible.

LuboVarga committed:
Make possible to consume single message in blocking way (like get command in http protocol is working, also wits saveAs equivalent).

Moved "execute next action" code from AmqpConsumeAction into AmqpConsumer.
Renamed ConsumerRequest to AsyncConsumerRequest and ConsumerRequest is now common trait also fow newly introduced ConsumeSingleMessageRequest.
consumeSingle dsl command was introduced.

LuboVarga committed:
Increased sbt version to more current.

LuboVarga committed:
Removed session attribute from AmqpConsumer class, as IMHO session should be passed as starting message for given scenario step (it is done so now).

LuboVarga committed:
Added runtime exception, when overwriting of builders internal state is detected. One example is, when there is call for consume after publish (publish would get overwritten without check).

LuboVarga committed:
Removed unnecessarily concrete map (HashMap) implementation from declaration of type Arguments. Also changed map from mutable to immutable version, as it is used in case class (immutable) and scala programmers should expect immutability also in its attributes.

LuboVarga committed:
Extended handling of precondition failed error also for queue declaration and binding declaration (I am not sure, if given error can happen also in binding).

LuboVarga committed:
Mainly added expression support for publish.

Added support for expressions to be used in publish message (i.e. message can be constructed from session attributes now).
Better handling of exceptions in single consume code.
Improved messages in precondition failed case of constructing exchanges/queues.

LuboVarga committed:
Added Engine.scala class from amqp archetype to be able to run simulations also from IDE.

LuboVarga committed:
Slightly modified ConsumingSimulation.scala example to show how to process (print) last consumed message.

LuboVarga committed:
Fixed ConsumingSimulation.scala example code.

LuboVarga committed:
Added simple script which I use to test if publishing/consuming works as in original project.

LuboVarga committed:
Changed maven configuration.

Now maven publishes also sources.
Added distribution management stub (just fill in your repository url in properties).

LuboVarga committed:
Extended readme with known limitations information and also with new possible consumeSingle operation.

Tiny refactoring in publisher.

LuboVarga committed:
Increased amqp client library version.

Maven version have defined tested range. In sbt there is latest version defined.

LuboVarga committed:
Redesigned consume single message code a bit.

Fixed also sample code and permissions for test script.

LuboVarga committed:
Optimized imports and improved maven build config to produce less warnings.

LuboVarga committed:
Added documentation for poolSize and simplified a bit code for creating pool of publishers.

Also removed needless watch for publishers.

LuboVarga committed:
Just backup commit. I have realized that AmqpConsumerCorrelation actor needs to be just one for one scenario step and not one per session/user.

Added correlation id to consume single request. It will make possible to consume specific message (specific reply for some published request).
Introduced AmqpConsumerBase class which holds primary termination things. It is common parent for AmqpConsumer and newly introduced AmqpConsumerCorrelation.
AmqpRouter has been extended to create AmqpConsumerCorrelation actor instead of AmqpConsumer for specific request type (with correlation id).

LuboVarga committed:
WIP First rpc support stub. Something in Publishing test is broken now.

Added new simulation test (works for now only with one parallel client due to design of current consumer.
AmqpActionBuilder can now build two different actions for ConsumeRequest (request with correlation id in mind have special action).
Nearly empty implementation of AmqpConsumeCorrelatedAction.
There were used queName parameter names instead exchangeName in PublishRequest class. This was fixed.
Also added new dsl command publishToQueue, which will make publish request to exchange with empty name and queue name will be used as routing key. This is way, how publishing directly to queue is possible through rabbitmq client library (have a look at RpcServer class for example).
Added missing declare exchange and bind to queue to consuming simulation. (this simulation is broken now)
Added support for new Rpc simulation to run script.

LuboVarga committed:
Fixed bug in code for consuming messages. Now Timeouts are correctly handled instead of NPE.

LuboVarga committed:
WIP Small progress in rpc support.

Make properties of message for publish also be expression (can be constructed using session).
Changed consume single code in such way, that it uses common (with consume all command) consumer, not special constructed only for one message.
Introduced new dsl command for publishing. It is rpcCall. It can take (or have default one) correlation id, which is filled into properties before publishing.
RpcSimulation echo implementation now takes correlation if from received message and copyies it into response.
Introduced user feeder into RpcSimulation. Now simulation also checks, that echo service has replyied with correct message body.

LuboVarga committed:
Added logback.xml config file to have console logging format a bit more readable.

LuboVarga committed:
WIP Implemented RPC support.

PublishRequest class was made trait and former class was replaced by PublishRequestAsync. Another child of PublishRequest is RpcCallRequest. Difference is that second mentioned starts next step of scenario a bit later then former one, because it saves correlationId (generated one, or provided one) to session. (have a look at AmqpPublishAction for additional insight)
Small refactorings.
AmqpPublishRequest now contains also next action needed for some specific cases. Should be refactored and moved. (TODO is in place of usage)
Added missing consumer start in consume single case in AmqpConsumer.
Also added saving stats to many places (some negative still missing).
Implemented AmqpConsumerCorrelation actor, which consumes just one message with requested correlationId.
Introduced publishRpcCall and consumeRpcResponse dsl commands.
Note, there are still some problem in some specific cases that simulation does not end despite that all work is done.

LuboVarga committed:
Fixed bug, where negative stats get saved also ako positive.

LuboVarga committed:
Make amqp request names visible in report (log them instead of static request type text only).

Added requestName to each request (publish/consume) and this name is used in title of operation.
requestName is by default (for all dsl builded requests) value from amqp request builder.
Added handling of unresolved consume with correlation requests during shutdown of actor.
Temporary added code for marking correlated consumptions of messages with statusCode header property failed, if it does not contain 200.

LuboVarga committed:
Fixed two compile errors.

LuboVarga committed:
Removed temporary UUID usage as correlation ID. Now session id and current time millis is used.

LuboVarga committed:
Implemented timeout (5 seconds now) for consuming message with correlation id. Timeouts are checked once per second.

Small cleanup of case classes.
Added statusCode = 200 to reply in RPC simulation, because it is checked in consumer right now by default (without any possibility to turn check off).

LuboVarga committed:
Added possibility to receive single message identified not only by correlation id, but by anything (user provides transforming function).

 Increased retry delay (if message arrives earlier than request is made) to 10 seconds.
 Status codes has been extended to accept any 2XX code as OK code and also missing code means that rpc response is OK.
 Introduced limit for logging (warn level) of unexpected messages (messages that was not requested).
 Small formatting improvements.

LuboVarga committed:
Removed debug printout.

LuboVarga committed:
Increased timeout for consuming single message from 5 to 60 seconds.

LuboVarga committed:
Merge branch 'consumeForSpecificCorrelationId'

LuboVarga committed:
WIP Migrate to gatling 2.2.2 dependency. Not compilable yet.

Primary change is from ActorRef type for actions to Action (ChainableAction) types.
Also used new amqp-client dependency.

LuboVarga committed:
Fixed maven configuration.

LuboVarga committed:
Increased artefact version and changed metadata of scm + developers.

LuboVarga committed:
Changed things about inheritance of logging traits to make project compilable again.

LuboVarga committed:
Fixed PathHelper import.

LuboVarga committed:
Moved setup of protocol specific things to proper place.

Removed stats engine from amqp thing. GatlingConfiguration holds it and I will use it soon.

LuboVarga committed:
Updated readme.

Commit:
Fixes to work with Gatling 2.2.2

LuboVarga committed:
Merge branch 'jpescalona-migratingToGatling2.2.2' into migratingToGatling2.2.2

LuboVarga committed:
Small changes in config to match my current deployment settings.

LuboVarga committed:
Small changes in config to match my current deployment settings.

LuboVarga committed:
Make release of version 0.7.

LuboVarga committed:
Merge branch '0.7'

LuboVarga committed:
New release with bugfix. Now more exchanges can be used in single scenario.

LuboVarga committed:
Fixed typo in versio.n

LuboVarga committed:
Fixing routing to right actor (AmqpConsume and AmqpConsumeCorrelation.

LuboVarga committed:
Map key have to contain also correlationConsumer info.

LuboVarga committed:
Going to 0.9 snapshot.

LuboVarga committed:
Fixed building consume request with correlations and some things around.

LuboVarga committed:
Releasing 0.9.
